### ISSUE TITLE
Add buttons for exporting plot information from visualization interface

### ISSUE THREAD

**Issue by fedarko** [biocore/qurro#87]
As discussed in #80. See comments there for implementation details w/r/t the sample plot.

Could also add a button for exporting rank plot information (it'd just be a list of the features included, I guess).

Code should detect if the currently used metadata field is `Sample ID` or `rankratioviz_balance`—if so, then the output data should only be two columns (since these fields would already be in the output anyway).

- [x] Export data from the sample plot
  - [x] Write code that tests the entirety of this functionality
    - Note to self: since `this.samplePlotJSON` is apparently modified by `changeSamplePlot()` (the balances demonstrably change, which is behavior that I was not aware was happening) we should be able to use samplePlotJSON instead of samplePlotView --> this will let us test this outside of #85
  - [x] ~~Make the sample plot export be predicated on the total number of samples shown (#92), not just whether or there are > 0 samples with a valid balance. Not super important, but would be nice.~~ **Update: actually, we're just gonna show all samples regardless of whether or not they're currently drawn. See #176.**
- [x] Export data from the rank plot about currently selected features
  - forum xref: https://forum.qiime2.org/t/qurro-export-numerator-denominator-list/13851
  - [x] Write tests for this

**Comment by ssteiche** [comment]
Very cool. I think your ideas discussed in #80 are spot on. The three values that I would love to have in a `.tsv` are Sample ID, current metadata value, and log ratio. I do wonder about how to arrange the currently selected feature IDs. If someone is filtering on a taxon the numerator and denominator could include very many features IDs that could be clunky dropped into a column header. This probably won't cause much trouble in a pandas df anyhow.

I would say maybe a second `.txt` with the feature IDs could work, but I'd hate to separate outputs like that if it isn't necessary.

**Comment by fedarko** [comment]
That's a good point re: the log ratio feature IDs. I don't think it'd break anything for pandas, but it might make things look ugly if there's a super-long column name for the log ratio.

One option would be putting all of the numerator feature IDs in the first line of the TSV file, all of the denominator feature IDs in the second line of the file, and then including all of the data afterwards (you could use something like the `skiprows` option in [`pandas.read_csv()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) to ignore these when reading the file).

However, I'm not so sure I like that option (having a normal TSV file would make parsing easier). I can probably get rankratioviz' visualization interface to output two files at once, one being a normal TSV for the sample plot (with just the log ratio column name being `feature_log_ratio` or something) and one being a list of feature IDs in the rank plot. This would work similarly to the screenshot functionality I just added in, where it tries to download the two files after you press a button once.

...That being said, this is thankfully a pretty minor problem—I'd imagine that once the initial functionality for this is ready, it'll be possible to change up the output details to be as useful as possible :)

**Comment by fedarko** [comment]
@ssteiche The latest version of rankratioviz should have controls to let you export a .tsv file of all currently-drawn points in the sample plot, with the three fields mentioned (sample ID, current log ratio, and current metadata value).

There are two caveats to this: 1) I don't have a solution for exporting the currently-selected ranked features ready yet, and 2) I haven't been able to write any unit tests to verify that the exporting functionality is working 100%. (To make a long story short, there's another issue (#85) which is causing me problems with testing certain parts of the visualization interface. I'll try to get to addressing that eventually.)

In the meantime, to fix 1) above—if you want to know which features have been selected, you can just copy the text in the big text boxes on the left of the display with all the feature IDs (the top box shows all selected "numerator" features, and the bottom box shows all selected "denominator" features).

Hope this helps!

**Comment by ssteiche** [comment]
This is working excellently, thanks again Marcus! I might need to add you as an author on the manuscript for all the time you've save me here :).

One thing that might make the output _slightly_ more portable would be to label the Log Ratio field in snake case `Log_Ratio`. I know it's silly, but it can make things really interface more smoothly. I read the `.tsv` into R to reproduce the plot with regression stats (below), but had a heck of a time figuring out what was throwing errors until I simply added in that underscore. Just one more example of how inexperienced users such as myself will butcher your program!

Thanks again, and this had been a tremendous help.

![image](https://user-images.githubusercontent.com/9448712/56834070-46a7ee00-6825-11e9-9a90-558a74995572.png)

**Comment by fedarko** [comment]
Haha I'm glad to hear things are working well! I'm super happy that this has been useful for your research.

Thanks for letting me know about the issue with spacing in the TSV file! The latest version of rankratioviz should solve the problems with the TSV file headers. `Sample ID` and `Log Ratio` are both now written as `Sample_ID` and `Log_Ratio`, and any metadata field values or sample IDs with whitespace in them are enclosed in double quotes (`""`). So the output should work well with most libraries (R included), hopefully! Again, thanks a ton for the heads up here—I'm glad I found out about this problem early so I could fix it up front :)

One more thing, speaking of problems: I recently realized that there was a bug in my code where samples with non-numeric values in the x-axis metadata field (e.g. `Missing: Not provided`) were being displayed on the chart with `x = 0`. Part of what led me to figuring this out was seeing [your screenshot yesterday](https://github.com/fedarko/rankratioviz/issues/80#issuecomment-486792702) in #80, where I noticed a sort of "clustering" of points around x = 0. I documented this issue in #91, and just fixed it in the latest version of rankratioviz. (Now, these points will be filtered from the chart based on whatever the current x-axis is, although they'll still be included in the exported TSV data.)

It looks like the chart you generated doesn't have this problem—I'm guessing R filtered out the non-numeric values. However, I figured I should let you know so you can update to the latest version (which doesn't exhibit this bug).

Thanks again for your suggestions and for using rankratioviz! :)

**Comment by ssteiche** [comment]
You are absolutely right! I was staring at the interactive plot for a bit and wondering to myself if I really was seeing a vertical line of points at zero. You're also correct that I filtered the null values out for the R plot, which seems like the appropriate decision so I'm glad rankratioviz will handle them this way now.
